moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (C) 2021 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi;
+
+import static java.lang.invoke.MethodType.methodType;
+
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.RecordComponent;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A {@link JsonAdapter} that supports Java {@code record} classes via reflection.
+ *
+ * <p><em>NOTE:</em> Java records require JDK 16 or higher.
+ */
+final class RecordJsonAdapter<T> extends JsonAdapter<T> {
+
+  static final JsonAdapter.Factory FACTORY =
+      (type, annotations, moshi) -> {
+        if (!annotations.isEmpty()) {
+          return null;
+        }
+
+        if (!(type instanceof Class) && !(type instanceof ParameterizedType)) {
+          return null;
+        }
+
+        var rawType = Types.getRawType(type);
+        if (!rawType.isRecord()) {
+          return null;
+        }
+
+        Map<String, Type> mappedTypeArgs = null;
+        if (type instanceof ParameterizedType parameterizedType) {
```

Comment:
This reads to me as very similar to `Util.resolve()`. Presumably we can’t use that here because this needs more records stuff? Either way this behavior really wants to be extracted into a separately-testable thing

Reply:
Fix: https://github.com/square/moshi/pull/1442/files